File: skylib-c++/src-lib/property.cpp

#include "property.hh"
#include "utils.hh"

namespace ch_skymarshall::gui {

using namespace std;

property::property(const string_view &_name, property_manager &_manager) :
		m_name(_name), m_manager(_manager) {
}

property::~property() {
	DESTR_LOG("~property " << m_name);
	m_manager.remove_listeners(m_name);
}

void property::attach() {
	m_attached = true;
}

const string_view& property::name() const {
	return m_name;
}

void property::add_listener(shared_ptr<property_listener> _listener) {
	m_manager.add_listener(m_name, _listener);
}

void property::remove_listener(weak_ptr<property_listener> _listener) {
	m_manager.remove_listener(m_name, _listener);
}

}

File: skylib-rust/src/lib_properties/property.rs
//! property.rs
//!
//! Rust port of `property.hh` + `property.cpp` (originally `Properties.hh`).
//!
//! Original author: scaille (Feb 19, 2012)
//!
//! ## Design notes
//!
//! C++'s `property` is an **abstract base class**: it owns the common
//! bookkeeping (name, owning `property_manager`, an `m_attached` flag) and
//! declares `attach()` pure virtual for subclasses to implement. Rust has
//! no inheritance, so this is split into two pieces, the way the rest of
//! this port handles the base-class-with-shared-state pattern:
//!
//! - [`Property`] — a trait capturing the abstract interface that other
//!   modules (`property_listener`, `property_manager`, `binding_chain`,
//!   ...) actually depend on.
//! - [`PropertyBase`] — a plain struct holding the fields `property.hh`
//!   declared (`m_name`, `m_manager`, `m_attached`), meant to be embedded
//!   by composition in concrete property types (see `typed_property.rs`),
//!   the way a C++ subclass would inherit them. Its `Drop` impl ports the
//!   `~property()` destructor, which deregisters all of this property's
//!   listeners from the manager.

use std::cell::Cell;
use std::rc::{Rc};
use crate::lib_properties::{PropertyManager, PropertyBase};


impl PropertyBase {
    pub fn new(name: &'static str, manager: Rc<PropertyManager>) -> Self {
        Self {
            name,
            manager,
            attached: Cell::new(false),
        }
    }
}

impl Drop for PropertyBase {
    fn drop(&mut self) {
        self.manager.remove_listeners(self.name);
    }
}
